SIGSEGV after update to 0.11.0 on listener

### Checks

- [x] I've already read https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/troubleshooting-actions-runner-controller-errors and I'm sure my issue is not covered in the troubleshooting guide.
- [x] I am using charts that are officially provided

### Controller Version

0.11.0

### Deployment Method

Helm

### Checks

- [x] This isn't a question or user support case (For Q&A and community support, go to [Discussions](https://github.com/actions/actions-runner-controller/discussions)).
- [x] I've read the [Changelog](https://github.com/actions/actions-runner-controller/blob/master/docs/gha-runner-scale-set-controller/README.md#changelog) before submitting this issue and I'm sure it's not due to any recently-introduced backward-incompatible changes

### To Reproduce

```markdown
1. Stay on 0.10.1 on controller and scale set
2. Upgrade to both 0.11.0
3. Check listener logs
4. Revert back to 0.10.1
```

### Describe the bug

During upgrade listener was recreated with the image version 0.11.0 and restarting due to error:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1844383]

goroutine 1 [running]:
github.com/actions/actions-runner-controller/cmd/ghalistener/app.New({{0xc0000465e0, 0x1a}, 0x0, 0x0, {0x0, 0x0}, {0xc000430120, 0x28}, {0xc0003053b0, 0x10}, ...})
    github.com/actions/actions-runner-controller/cmd/ghalistener/app/app.go:73 +0x3e3
main.main()
    github.com/actions/actions-runner-controller/cmd/ghalistener/main.go:27 +0x1f8
Stream closed EOF for gha-runners-controller/k8s-dind-6746d74f-listener (listener)
```
Reverting back to 0.10.1 fixes the issue.
Config wasn't changed during the upgrade.

PS. Controller has no backward compatibility for 0.11.0 version with scale set on 0.10.1. Version should be aligned.

### Describe the expected behavior

Smooth upgrade from 0.10.1 to 0.11.0

### Additional Context

```yaml
Ubuntu 24.04.1 LTS (GNU/Linux 6.8.0-45-generic x86_64)
k8s version: v1.31.1+rke2r1
```

### Controller Logs

```shell
Not provided
```

### Runner Pod Logs

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1844383]

goroutine 1 [running]:
github.com/actions/actions-runner-controller/cmd/ghalistener/app.New({{0xc0000465e0, 0x1a}, 0x0, 0x0, {0x0, 0x0}, {0xc000430120, 0x28}, {0xc0003053b0, 0x10}, ...})
    github.com/actions/actions-runner-controller/cmd/ghalistener/app/app.go:73 +0x3e3
main.main()
    github.com/actions/actions-runner-controller/cmd/ghalistener/main.go:27 +0x1f8
Stream closed EOF for gha-runners-controller/k8s-dind-6746d74f-listener (listener)
```

Same error's popping up, even with `listenerMetrics` set

Error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1844383]

goroutine 1 [running]:
github.com/actions/actions-runner-controller/cmd/ghalistener/app.New({{0xc00048b1a0, 0x2e}, 0x0, 0x0, {0x0, 0x0}, {0xc00048b1d0, 0x28}, {0xc000158858, 0x11}, ...})
	github.com/actions/actions-runner-controller/cmd/ghalistener/app/app.go:73 +0x3e3
main.main()
	github.com/actions/actions-runner-controller/cmd/ghalistener/main.go:27 +0x1f8
```

--- a/cmd/ghalistener/app/app.go
+++ b/cmd/ghalistener/app/app.go
@@ -61,7 +61,7 @@ func New(config config.Config) (*App, error) {
 	}
 
 	if config.MetricsAddr != "" {
-		exporterConfig := metrics.ExporterConfig{
+		app.metrics = metrics.NewExporter(metrics.ExporterConfig{
 			ScaleSetName:      config.EphemeralRunnerSetName,
 			ScaleSetNamespace: config.EphemeralRunnerSetNamespace,
 			Enterprise:        ghConfig.Enterprise,
@@ -69,14 +69,9 @@ func New(config config.Config) (*App, error) {
 			Repository:        ghConfig.Repository,
 			ServerAddr:        config.MetricsAddr,
 			ServerEndpoint:    config.MetricsEndpoint,
+			Metrics:           config.Metrics,
 			Logger:            app.logger.WithName("metrics exporter"),
-		}
-
-		if config.Metrics != nil {
-			exporterConfig.Metrics = *config.Metrics
-		}
-
-		app.metrics = metrics.NewExporter(exporterConfig)
+		})
 	}
 
 	worker, err := worker.New(

--- a/cmd/ghalistener/app/app.go
+++ b/cmd/ghalistener/app/app.go
@@ -61,7 +61,7 @@ func New(config config.Config) (*App, error) {
 	}
 
 	if config.MetricsAddr != "" {
-		app.metrics = metrics.NewExporter(metrics.ExporterConfig{
+		exporterConfig := metrics.ExporterConfig{
 			ScaleSetName:      config.EphemeralRunnerSetName,
 			ScaleSetNamespace: config.EphemeralRunnerSetNamespace,
 			Enterprise:        ghConfig.Enterprise,
@@ -70,8 +70,13 @@ func New(config config.Config) (*App, error) {
 			ServerAddr:        config.MetricsAddr,
 			ServerEndpoint:    config.MetricsEndpoint,
 			Logger:            app.logger.WithName("metrics exporter"),
-			Metrics:           *config.Metrics,
-		})
+		}
+
+		if config.Metrics != nil {
+			exporterConfig.Metrics = *config.Metrics
+		}
+
+		app.metrics = metrics.NewExporter(exporterConfig)
 	}
 
 	worker, err := worker.New(

--- a/cmd/ghalistener/metrics/metrics.go
+++ b/cmd/ghalistener/metrics/metrics.go
@@ -155,134 +155,137 @@ type ExporterConfig struct {
 	Metrics           *v1alpha1.MetricsConfig
 }
 
+var defaultMetrics = v1alpha1.MetricsConfig{
+	Counters: map[string]*v1alpha1.CounterMetric{
+		MetricStartedJobsTotal: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyJobName,
+				labelKeyEventName,
+			},
+		},
+		MetricCompletedJobsTotal: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyJobName,
+				labelKeyEventName,
+				labelKeyJobResult,
+			},
+		},
+	},
+	Gauges: map[string]*v1alpha1.GaugeMetric{
+		MetricAssignedJobs: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyRunnerScaleSetName,
+				labelKeyRunnerScaleSetNamespace,
+			},
+		},
+		MetricRunningJobs: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyRunnerScaleSetName,
+				labelKeyRunnerScaleSetNamespace,
+			},
+		},
+		MetricRegisteredRunners: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyRunnerScaleSetName,
+				labelKeyRunnerScaleSetNamespace,
+			},
+		},
+		MetricBusyRunners: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyRunnerScaleSetName,
+				labelKeyRunnerScaleSetNamespace,
+			},
+		},
+		MetricMinRunners: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyRunnerScaleSetName,
+				labelKeyRunnerScaleSetNamespace,
+			},
+		},
+		MetricMaxRunners: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyRunnerScaleSetName,
+				labelKeyRunnerScaleSetNamespace,
+			},
+		},
+		MetricDesiredRunners: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyRunnerScaleSetName,
+				labelKeyRunnerScaleSetNamespace,
+			},
+		},
+		MetricIdleRunners: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyRunnerScaleSetName,
+				labelKeyRunnerScaleSetNamespace,
+			},
+		},
+	},
+	Histograms: map[string]*v1alpha1.HistogramMetric{
+		MetricJobStartupDurationSeconds: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyJobName,
+				labelKeyEventName,
+			},
+			Buckets: defaultRuntimeBuckets,
+		},
+		MetricJobExecutionDurationSeconds: {
+			Labels: []string{
+				labelKeyEnterprise,
+				labelKeyOrganization,
+				labelKeyRepository,
+				labelKeyJobName,
+				labelKeyEventName,
+				labelKeyJobResult,
+			},
+			Buckets: defaultRuntimeBuckets,
+		},
+	},
+}
+
 func (e *ExporterConfig) defaults() {
 	if e.ServerAddr == "" {
-		e.Metrics = nil
-		return
+		e.ServerAddr = ":8080"
 	}
-	if e.Metrics != nil {
-		return
+	if e.ServerEndpoint == "" {
+		e.ServerEndpoint = "/metrics"
 	}
-
-	e.Metrics = &v1alpha1.MetricsConfig{
-		Counters: map[string]*v1alpha1.CounterMetric{
-			MetricStartedJobsTotal: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyJobName,
-					labelKeyEventName,
-				},
-			},
-			MetricCompletedJobsTotal: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyJobName,
-					labelKeyEventName,
-					labelKeyJobResult,
-				},
-			},
-		},
-		Gauges: map[string]*v1alpha1.GaugeMetric{
-			MetricAssignedJobs: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyRunnerScaleSetName,
-					labelKeyRunnerScaleSetNamespace,
-				},
-			},
-			MetricRunningJobs: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyRunnerScaleSetName,
-					labelKeyRunnerScaleSetNamespace,
-				},
-			},
-			MetricRegisteredRunners: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyRunnerScaleSetName,
-					labelKeyRunnerScaleSetNamespace,
-				},
-			},
-			MetricBusyRunners: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyRunnerScaleSetName,
-					labelKeyRunnerScaleSetNamespace,
-				},
-			},
-			MetricMinRunners: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyRunnerScaleSetName,
-					labelKeyRunnerScaleSetNamespace,
-				},
-			},
-			MetricMaxRunners: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyRunnerScaleSetName,
-					labelKeyRunnerScaleSetNamespace,
-				},
-			},
-			MetricDesiredRunners: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyRunnerScaleSetName,
-					labelKeyRunnerScaleSetNamespace,
-				},
-			},
-			MetricIdleRunners: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyRunnerScaleSetName,
-					labelKeyRunnerScaleSetNamespace,
-				},
-			},
-		},
-		Histograms: map[string]*v1alpha1.HistogramMetric{
-			MetricJobStartupDurationSeconds: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyJobName,
-					labelKeyEventName,
-				},
-				Buckets: defaultRuntimeBuckets,
-			},
-			MetricJobExecutionDurationSeconds: {
-				Labels: []string{
-					labelKeyEnterprise,
-					labelKeyOrganization,
-					labelKeyRepository,
-					labelKeyJobName,
-					labelKeyEventName,
-					labelKeyJobResult,
-				},
-				Buckets: defaultRuntimeBuckets,
-			},
-		},
+	if e.Metrics == nil {
+		defaultMetrics := defaultMetrics
+		e.Metrics = &defaultMetrics
 	}
 }
 

--- a/cmd/ghalistener/metrics/metrics.go
+++ b/cmd/ghalistener/metrics/metrics.go
@@ -152,13 +152,145 @@ type ExporterConfig struct {
 	ServerAddr        string
 	ServerEndpoint    string
 	Logger            logr.Logger
-	Metrics           v1alpha1.MetricsConfig
+	Metrics           *v1alpha1.MetricsConfig
+}
+
+func (e *ExporterConfig) defaults() {
+	if e.ServerAddr == "" {
+		e.Metrics = nil
+		return
+	}
+	if e.Metrics != nil {
+		return
+	}
+
+	e.Metrics = &v1alpha1.MetricsConfig{
+		Counters: map[string]*v1alpha1.CounterMetric{
+			MetricStartedJobsTotal: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyJobName,
+					labelKeyEventName,
+				},
+			},
+			MetricCompletedJobsTotal: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyJobName,
+					labelKeyEventName,
+					labelKeyJobResult,
+				},
+			},
+		},
+		Gauges: map[string]*v1alpha1.GaugeMetric{
+			MetricAssignedJobs: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyRunnerScaleSetName,
+					labelKeyRunnerScaleSetNamespace,
+				},
+			},
+			MetricRunningJobs: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyRunnerScaleSetName,
+					labelKeyRunnerScaleSetNamespace,
+				},
+			},
+			MetricRegisteredRunners: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyRunnerScaleSetName,
+					labelKeyRunnerScaleSetNamespace,
+				},
+			},
+			MetricBusyRunners: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyRunnerScaleSetName,
+					labelKeyRunnerScaleSetNamespace,
+				},
+			},
+			MetricMinRunners: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyRunnerScaleSetName,
+					labelKeyRunnerScaleSetNamespace,
+				},
+			},
+			MetricMaxRunners: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyRunnerScaleSetName,
+					labelKeyRunnerScaleSetNamespace,
+				},
+			},
+			MetricDesiredRunners: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyRunnerScaleSetName,
+					labelKeyRunnerScaleSetNamespace,
+				},
+			},
+			MetricIdleRunners: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyRunnerScaleSetName,
+					labelKeyRunnerScaleSetNamespace,
+				},
+			},
+		},
+		Histograms: map[string]*v1alpha1.HistogramMetric{
+			MetricJobStartupDurationSeconds: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyJobName,
+					labelKeyEventName,
+				},
+				Buckets: defaultRuntimeBuckets,
+			},
+			MetricJobExecutionDurationSeconds: {
+				Labels: []string{
+					labelKeyEnterprise,
+					labelKeyOrganization,
+					labelKeyRepository,
+					labelKeyJobName,
+					labelKeyEventName,
+					labelKeyJobResult,
+				},
+				Buckets: defaultRuntimeBuckets,
+			},
+		},
+	}
 }
 
 func NewExporter(config ExporterConfig) ServerExporter {
+	config.defaults()
 	reg := prometheus.NewRegistry()
 
-	metrics := installMetrics(config.Metrics, reg, config.Logger)
+	metrics := installMetrics(*config.Metrics, reg, config.Logger)
 
 	mux := http.NewServeMux()
 	mux.Handle(

--- a/cmd/ghalistener/metrics/metrics_test.go
+++ b/cmd/ghalistener/metrics/metrics_test.go
@@ -109,22 +109,22 @@ func TestNewExporter(t *testing.T) {
 		reg := prometheus.NewRegistry()
 		wantMetrics := installMetrics(defaultMetrics, reg, config.Logger)
 
-		assert.Equal(t, len(wantMetrics.counters), len(exporter.metrics.counters))
+		assert.Equal(t, len(wantMetrics.counters), len(exporter.counters))
 		for k, v := range wantMetrics.counters {
-			assert.Contains(t, exporter.metrics.counters, k)
-			assert.Equal(t, v.config, exporter.metrics.counters[k].config)
+			assert.Contains(t, exporter.counters, k)
+			assert.Equal(t, v.config, exporter.counters[k].config)
 		}
 
-		assert.Equal(t, len(wantMetrics.gauges), len(exporter.metrics.gauges))
+		assert.Equal(t, len(wantMetrics.gauges), len(exporter.gauges))
 		for k, v := range wantMetrics.gauges {
-			assert.Contains(t, exporter.metrics.gauges, k)
-			assert.Equal(t, v.config, exporter.metrics.gauges[k].config)
+			assert.Contains(t, exporter.gauges, k)
+			assert.Equal(t, v.config, exporter.gauges[k].config)
 		}
 
-		assert.Equal(t, len(wantMetrics.histograms), len(exporter.metrics.histograms))
+		assert.Equal(t, len(wantMetrics.histograms), len(exporter.histograms))
 		for k, v := range wantMetrics.histograms {
-			assert.Contains(t, exporter.metrics.histograms, k)
-			assert.Equal(t, v.config, exporter.metrics.histograms[k].config)
+			assert.Contains(t, exporter.histograms, k)
+			assert.Equal(t, v.config, exporter.histograms[k].config)
 		}
 
 		require.NotNil(t, exporter.srv)
@@ -151,22 +151,22 @@ func TestNewExporter(t *testing.T) {
 		reg := prometheus.NewRegistry()
 		wantMetrics := installMetrics(defaultMetrics, reg, config.Logger)
 
-		assert.Equal(t, len(wantMetrics.counters), len(exporter.metrics.counters))
+		assert.Equal(t, len(wantMetrics.counters), len(exporter.counters))
 		for k, v := range wantMetrics.counters {
-			assert.Contains(t, exporter.metrics.counters, k)
-			assert.Equal(t, v.config, exporter.metrics.counters[k].config)
+			assert.Contains(t, exporter.counters, k)
+			assert.Equal(t, v.config, exporter.counters[k].config)
 		}
 
-		assert.Equal(t, len(wantMetrics.gauges), len(exporter.metrics.gauges))
+		assert.Equal(t, len(wantMetrics.gauges), len(exporter.gauges))
 		for k, v := range wantMetrics.gauges {
-			assert.Contains(t, exporter.metrics.gauges, k)
-			assert.Equal(t, v.config, exporter.metrics.gauges[k].config)
+			assert.Contains(t, exporter.gauges, k)
+			assert.Equal(t, v.config, exporter.gauges[k].config)
 		}
 
-		assert.Equal(t, len(wantMetrics.histograms), len(exporter.metrics.histograms))
+		assert.Equal(t, len(wantMetrics.histograms), len(exporter.histograms))
 		for k, v := range wantMetrics.histograms {
-			assert.Contains(t, exporter.metrics.histograms, k)
-			assert.Equal(t, v.config, exporter.metrics.histograms[k].config)
+			assert.Contains(t, exporter.histograms, k)
+			assert.Equal(t, v.config, exporter.histograms[k].config)
 		}
 
 		require.NotNil(t, exporter.srv)
@@ -212,22 +212,22 @@ func TestNewExporter(t *testing.T) {
 		reg := prometheus.NewRegistry()
 		wantMetrics := installMetrics(metricsConfig, reg, config.Logger)
 
-		assert.Equal(t, len(wantMetrics.counters), len(exporter.metrics.counters))
+		assert.Equal(t, len(wantMetrics.counters), len(exporter.counters))
 		for k, v := range wantMetrics.counters {
-			assert.Contains(t, exporter.metrics.counters, k)
-			assert.Equal(t, v.config, exporter.metrics.counters[k].config)
+			assert.Contains(t, exporter.counters, k)
+			assert.Equal(t, v.config, exporter.counters[k].config)
 		}
 
-		assert.Equal(t, len(wantMetrics.gauges), len(exporter.metrics.gauges))
+		assert.Equal(t, len(wantMetrics.gauges), len(exporter.gauges))
 		for k, v := range wantMetrics.gauges {
-			assert.Contains(t, exporter.metrics.gauges, k)
-			assert.Equal(t, v.config, exporter.metrics.gauges[k].config)
+			assert.Contains(t, exporter.gauges, k)
+			assert.Equal(t, v.config, exporter.gauges[k].config)
 		}
 
-		assert.Equal(t, len(wantMetrics.histograms), len(exporter.metrics.histograms))
+		assert.Equal(t, len(wantMetrics.histograms), len(exporter.histograms))
 		for k, v := range wantMetrics.histograms {
-			assert.Contains(t, exporter.metrics.histograms, k)
-			assert.Equal(t, v.config, exporter.metrics.histograms[k].config)
+			assert.Contains(t, exporter.histograms, k)
+			assert.Equal(t, v.config, exporter.histograms[k].config)
 		}
 
 		require.NotNil(t, exporter.srv)

--- a/cmd/ghalistener/metrics/metrics_test.go
+++ b/cmd/ghalistener/metrics/metrics_test.go
@@ -7,6 +7,7 @@ import (
 	"github.com/go-logr/logr"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestInstallMetrics(t *testing.T) {
@@ -85,4 +86,180 @@ func TestInstallMetrics(t *testing.T) {
 	duration := got.histograms[MetricJobStartupDurationSeconds]
 	assert.Equal(t, duration.config.Labels, metricsConfig.Histograms[MetricJobStartupDurationSeconds].Labels)
 	assert.Equal(t, duration.config.Buckets, defaultRuntimeBuckets)
+}
+
+func TestNewExporter(t *testing.T) {
+	t.Run("with defaults metrics applied", func(t *testing.T) {
+		config := ExporterConfig{
+			ScaleSetName:      "test-scale-set",
+			ScaleSetNamespace: "test-namespace",
+			Enterprise:        "",
+			Organization:      "org",
+			Repository:        "repo",
+			ServerAddr:        ":6060",
+			ServerEndpoint:    "/metrics",
+			Logger:            logr.Discard(),
+			Metrics:           nil, // when metrics is nil, all default metrics should be registered
+		}
+
+		exporter, ok := NewExporter(config).(*exporter)
+		require.True(t, ok, "expected exporter to be of type *exporter")
+		require.NotNil(t, exporter)
+
+		reg := prometheus.NewRegistry()
+		wantMetrics := installMetrics(defaultMetrics, reg, config.Logger)
+
+		assert.Equal(t, len(wantMetrics.counters), len(exporter.metrics.counters))
+		for k, v := range wantMetrics.counters {
+			assert.Contains(t, exporter.metrics.counters, k)
+			assert.Equal(t, v.config, exporter.metrics.counters[k].config)
+		}
+
+		assert.Equal(t, len(wantMetrics.gauges), len(exporter.metrics.gauges))
+		for k, v := range wantMetrics.gauges {
+			assert.Contains(t, exporter.metrics.gauges, k)
+			assert.Equal(t, v.config, exporter.metrics.gauges[k].config)
+		}
+
+		assert.Equal(t, len(wantMetrics.histograms), len(exporter.metrics.histograms))
+		for k, v := range wantMetrics.histograms {
+			assert.Contains(t, exporter.metrics.histograms, k)
+			assert.Equal(t, v.config, exporter.metrics.histograms[k].config)
+		}
+
+		require.NotNil(t, exporter.srv)
+		assert.Equal(t, config.ServerAddr, exporter.srv.Addr)
+	})
+
+	t.Run("with default server URL", func(t *testing.T) {
+		config := ExporterConfig{
+			ScaleSetName:      "test-scale-set",
+			ScaleSetNamespace: "test-namespace",
+			Enterprise:        "",
+			Organization:      "org",
+			Repository:        "repo",
+			ServerAddr:        "", // empty ServerAddr should default to ":8080"
+			ServerEndpoint:    "",
+			Logger:            logr.Discard(),
+			Metrics:           nil, // when metrics is nil, all default metrics should be registered
+		}
+
+		exporter, ok := NewExporter(config).(*exporter)
+		require.True(t, ok, "expected exporter to be of type *exporter")
+		require.NotNil(t, exporter)
+
+		reg := prometheus.NewRegistry()
+		wantMetrics := installMetrics(defaultMetrics, reg, config.Logger)
+
+		assert.Equal(t, len(wantMetrics.counters), len(exporter.metrics.counters))
+		for k, v := range wantMetrics.counters {
+			assert.Contains(t, exporter.metrics.counters, k)
+			assert.Equal(t, v.config, exporter.metrics.counters[k].config)
+		}
+
+		assert.Equal(t, len(wantMetrics.gauges), len(exporter.metrics.gauges))
+		for k, v := range wantMetrics.gauges {
+			assert.Contains(t, exporter.metrics.gauges, k)
+			assert.Equal(t, v.config, exporter.metrics.gauges[k].config)
+		}
+
+		assert.Equal(t, len(wantMetrics.histograms), len(exporter.metrics.histograms))
+		for k, v := range wantMetrics.histograms {
+			assert.Contains(t, exporter.metrics.histograms, k)
+			assert.Equal(t, v.config, exporter.metrics.histograms[k].config)
+		}
+
+		require.NotNil(t, exporter.srv)
+		assert.Equal(t, exporter.srv.Addr, ":8080")
+	})
+
+	t.Run("with metrics configured", func(t *testing.T) {
+		metricsConfig := v1alpha1.MetricsConfig{
+			Counters: map[string]*v1alpha1.CounterMetric{
+				MetricStartedJobsTotal: {
+					Labels: []string{labelKeyRepository},
+				},
+			},
+			Gauges: map[string]*v1alpha1.GaugeMetric{
+				MetricAssignedJobs: {
+					Labels: []string{labelKeyRepository},
+				},
+			},
+			Histograms: map[string]*v1alpha1.HistogramMetric{
+				MetricJobExecutionDurationSeconds: {
+					Labels:  []string{labelKeyRepository},
+					Buckets: []float64{0.1, 1},
+				},
+			},
+		}
+
+		config := ExporterConfig{
+			ScaleSetName:      "test-scale-set",
+			ScaleSetNamespace: "test-namespace",
+			Enterprise:        "",
+			Organization:      "org",
+			Repository:        "repo",
+			ServerAddr:        ":6060",
+			ServerEndpoint:    "/metrics",
+			Logger:            logr.Discard(),
+			Metrics:           &metricsConfig,
+		}
+
+		exporter, ok := NewExporter(config).(*exporter)
+		require.True(t, ok, "expected exporter to be of type *exporter")
+		require.NotNil(t, exporter)
+
+		reg := prometheus.NewRegistry()
+		wantMetrics := installMetrics(metricsConfig, reg, config.Logger)
+
+		assert.Equal(t, len(wantMetrics.counters), len(exporter.metrics.counters))
+		for k, v := range wantMetrics.counters {
+			assert.Contains(t, exporter.metrics.counters, k)
+			assert.Equal(t, v.config, exporter.metrics.counters[k].config)
+		}
+
+		assert.Equal(t, len(wantMetrics.gauges), len(exporter.metrics.gauges))
+		for k, v := range wantMetrics.gauges {
+			assert.Contains(t, exporter.metrics.gauges, k)
+			assert.Equal(t, v.config, exporter.metrics.gauges[k].config)
+		}
+
+		assert.Equal(t, len(wantMetrics.histograms), len(exporter.metrics.histograms))
+		for k, v := range wantMetrics.histograms {
+			assert.Contains(t, exporter.metrics.histograms, k)
+			assert.Equal(t, v.config, exporter.metrics.histograms[k].config)
+		}
+
+		require.NotNil(t, exporter.srv)
+		assert.Equal(t, config.ServerAddr, exporter.srv.Addr)
+	})
+}
+
+func TestExporterConfigDefaults(t *testing.T) {
+	config := ExporterConfig{
+		ScaleSetName:      "test-scale-set",
+		ScaleSetNamespace: "test-namespace",
+		Enterprise:        "",
+		Organization:      "org",
+		Repository:        "repo",
+		ServerAddr:        "",
+		ServerEndpoint:    "",
+		Logger:            logr.Discard(),
+		Metrics:           nil, // when metrics is nil, all default metrics should be registered
+	}
+
+	config.defaults()
+	want := ExporterConfig{
+		ScaleSetName:      "test-scale-set",
+		ScaleSetNamespace: "test-namespace",
+		Enterprise:        "",
+		Organization:      "org",
+		Repository:        "repo",
+		ServerAddr:        ":8080",    // default server address
+		ServerEndpoint:    "/metrics", // default server endpoint
+		Logger:            logr.Discard(),
+		Metrics:           &defaultMetrics, // when metrics is nil, all default metrics should be registered
+	}
+
+	assert.Equal(t, want, config)
 }

